src/Monotonic.cpp

```diff
@@ -30,26 +30,172 @@ using std::string;
+    return a.is_single_point(0);
+}
+
+bool is_monotonic_increasing(const ConstantInterval &a) {
+    return a.has_lower_bound() && a.min >= 0;
+}
+
+bool is_monotonic_decreasing(const ConstantInterval &a) {
+    return a.has_upper_bound() && a.max <= 0;
+}
+
+ConstantInterval to_interval(Monotonic m) {
+    switch (m) {
+    case Monotonic::Constant:
+        return ConstantInterval::single_point(0);
+    case Monotonic::Increasing:
+        return ConstantInterval::bounded_below(0);
+    case Monotonic::Decreasing:
+        return ConstantInterval::bounded_above(0);
+    case Monotonic::Unknown:
+        return ConstantInterval();
+    }
+    return ConstantInterval();
+}
+
+Monotonic to_monotonic(const ConstantInterval &x) {
+    if (is_constant(x)) {
+        return Monotonic::Constant;
+    } else if (is_monotonic_increasing(x)) {
+        return Monotonic::Increasing;
+    } else if (is_monotonic_decreasing(x)) {
+        return Monotonic::Decreasing;
+    } else {
+        return Monotonic::Unknown;
+    }
+}
+
+ConstantInterval unify(const ConstantInterval &a, const ConstantInterval &b) {
+    return ConstantInterval::make_union(a, b);
+}
+
+ConstantInterval unify(const ConstantInterval &a, int64_t b) {
+    ConstantInterval result;
+    result.include(b);
+    return result;
+}
+
+// Helpers for doing arithmetic on ConstantIntervals that avoid generating
+// expressions of pos_inf/neg_inf.
+ConstantInterval add(const ConstantInterval &a, const ConstantInterval &b) {
+    ConstantInterval result;
+    result.min_defined = a.has_lower_bound() && b.has_lower_bound();
+    result.max_defined = a.has_upper_bound() && b.has_upper_bound();
+    if (result.has_lower_bound()) {
+        result.min = a.min + b.min;
+    }
+    if (result.has_upper_bound()) {
+        result.max = a.max + b.max;
+    }
+    return result;
+}
+
+ConstantInterval add(const ConstantInterval &a, int64_t b) {
+    return add(a, ConstantInterval(b, b));
+}
+
+ConstantInterval negate(const ConstantInterval &r) {
+    ConstantInterval result;
+    result.min_defined = r.has_upper_bound();
+    result.min = r.has_upper_bound() ? -r.max : 0;
+    result.max_defined = r.has_lower_bound();
+    result.max = r.has_lower_bound() ? -r.min : 0;
+    return result;
+}
+
+ConstantInterval sub(const ConstantInterval &a, const ConstantInterval &b) {
+    return add(a, negate(b));
+}
+
+ConstantInterval sub(const ConstantInterval &a, int64_t b) {
+    return sub(a, ConstantInterval(b, b));
+}
+
+ConstantInterval multiply(const ConstantInterval &a, int64_t b) {
+    ConstantInterval result(a);
+    if (b < 0) {
+        result = negate(result);
+        b = -b;
+    }
+    if (result.has_lower_bound()) {
+        result.min *= b;
+    }
+    if (result.has_upper_bound()) {
+        result.max *= b;
+    }
+    return result;
+}
+
+ConstantInterval multiply(const ConstantInterval &a, const ConstantInterval &b) {
+    std::vector<int64_t> bounds;
```

Comment:
I wonder if the heap allocation here is a bottleneck. This seems like it could be faster with a stack array and a pointer to the end element. (e.g. *end++ = a.min * b.max)